docs/proposal/2022-11-22-botkube-plugins-integration.md

```diff
@@ -0,0 +1,498 @@
+          "required": [
+              "version"
+          ],
+          "properties": {
+            "version": {
+              "$id": "#/properties/version",
+              "type": "string",
+              "minLength": 5,
+              "pattern": "^(?:0|[1-9]\\d*)\\.(?:0|[1-9]\\d*)\\.(?:0|[1-9]\\d*)$",
+              "title": "Kubernetes version",
+              "description": "Kubernetes version",
+              "default": "",
+              "examples": [
+                  "1.19.0"
+              ]
+            }
+          },
+          "additionalProperties": true
+        }
+```
+
+This will be used by Botkube to validate that configuration defined by user is valid before even starting a given plugin.
+
+### Passing configuration to a given plugin
+
+Current Botkube implementation allows you to specify different executor and sources configuration and bind them to a single channel. Later they are merged together based on the given business logic.
+
+To support the same experience for plugins, we need to pass a list of configuration each time we run the:
+
+- `Execute` method for executor plugins
+- `Source` method for source plugins
+
+We pass an array of strings, so later they can be unmarshalled by a given plugin and merged based on custom business logic.
+
+We don't support configuration validation inside Botkube. See validating [plugins configuration](#plugin-configuration) for more information.
+
+The allowed configuration parameters will be described on the `docs.botkube.io` site in the same way that we did for `kubectl` executor and `kuberneetes` source.
+
+### Refreshing already loaded plugins
+
+Once the Botkube starts it caches:
+
+- all index files for defined plugins repositories
+- all enabled plugins binaries
+
+We use the `emptyDir` so the data is not removed as long as the Pod is not rescheduled to a different Node. To force download, you need to run:
+
+```bash
+kubectl exec -it $(kubectl get po -l app=botkube -n botkube -oname) -- rm -rf /tmp
+kubectl delete po -l app=botkube -n botkube
+```
+
+Later we can provide a dedicated `@Botkube` command to simplify refreshing downloaded plugins. However, in the happy path scenarios no one should replace already release binaries.
+
+### Supporting multi OS and architectures
+
+We want to run plugins on different Kubernetes distros and also locally for development purposes. As a result, we need to support different operating systems and different architectures.
+
+We enforce that plugin binaries have a suffix with a given pattern: `{os}_{arch}[.exe]`. For example:
+
+- `botkube_executor_helm_linux_amd64`
+- `botkube_source_prometheus_darwin_amd64`
+
+Botkube uses the `runtime.GOOS` and `runtime.GOARCH` variables to determine on which system Botkube is running. It downloads respected binary by searching a proper link in plugin index file:
+
+```yaml
+entries:
+  - name: "kubectl"
+    type: "executor"
+    description: "Kubectl executor plugin."
+    version: "v1.0.0"
+    links:
```

Comment:
BTW (minor) I like `urls` more - this is what Helm has 😛 Or `binaries`?

Ignore this comment, if you don't like both options 😄